src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
+        {
+            _schema = schema;
+            _typeToPropertyValueConstructorMap = primitiveValueBuilders ?? GetBuildersForRequiredPrimitives();
+        }
+
+        private JsonSchema _schema;
+        private IDictionary<Type, PrimitiveValueBuilder> _typeToPropertyValueConstructorMap;
+
+        // There are two places in the format where object nest references to the defining 
+        // type. 'exception.innerExceptions' is an array of exception objects. 
+        // 'node.children' is another collection of nodes. We need to watch for these
+        // scenarios to prevent re-entrancy and eventual stack consumption when 
+        // producing samples of these types.
+        private int graphNodeParentCount;
+        private int exceptionDataParentCount;
+
+        public override object VisitActual(ISarifNode node)
+        {
+            PopulateInstanceWithDefaultMemberValues(node);
+
+            return base.VisitActual(node);
+        }
+
+        // Retain nesting level for visiting exceptions to prevent
+        // unbounded re-entrance populating exception.innerExceptions
+        public override ExceptionData VisitExceptionData(ExceptionData node)
+        {
+            exceptionDataParentCount++;
+            node = base.VisitExceptionData(node);
+            exceptionDataParentCount--;
+
+            return node;
+        }
+
+        // Retain nesting level for visiting graph nodes, to prevent
+        // unbounded re-entrance populating exception.innerExceptions
+        public override Node VisitNode(Node node)
+        {
+            graphNodeParentCount++;
+            node = base.VisitNode(node);
+            graphNodeParentCount--;
+
+            return node;
+        }
+
+        private void PopulateInstanceWithDefaultMemberValues(ISarifNode node)        {
+           
+            Type nodeType = node.GetType();
+            var binding = BindingFlags.Public | BindingFlags.Instance;
+            foreach (PropertyInfo property in nodeType.GetProperties(binding))
+            {
+                // The node kind is always properly set in the OM and
+                // isn't relevant to the SARIF schema itself
+                if (property.Name == "SarifNodeKind") { continue; }
+
+                // Property bags and tags are populated via special methods on 
+                // the class rather than direct access of properties.
```

Comment:
>. [](start = 68, length = 1)

Worth a comment: `PropertyNames` and `Tags` are inherited from `PropertyBagHolder`.